File: api/src/docs/member.py
from beanie import PydanticObjectId

from src.models.member import MemberModel

from .base import Example, response


member_response = response(
    description='Member Found',
    model=MemberModel,
    examples=[
        Example(
            name='Member with Userproxy (No Token Scope)',
            value=MemberModel(
                id=PydanticObjectId('67cab2a55f8b2e7fd3d27d0c'),
                name='steve',
                meta='',
                pronouns='any/all',
                bio='example bio',
                birthday='01/01/2000',
                color=16777065,
                avatar='f873c5cff5608ff4952cfe8b80f1a86c',
                proxy_tags=[
                    MemberModel.ProxyTag(
                        id=PydanticObjectId('67cab28fe3188a0a7807c8fd'),
                        prefix='s;',
                        suffix='',
                        regex=False,
                        case_sensitive=False,
                        avatar=None),
                    MemberModel.ProxyTag(
                        id=PydanticObjectId('67cab28fe3188a0a7807c8fe'),
                        prefix='',
                        suffix='--steve',
                        regex=False,
                        case_sensitive=False,
                        avatar=None)],
                userproxy=MemberModel.UserProxy(
                    bot_id='1297857704512978984',
                    public_key='',
                    token='',
                    command='st',
                    guilds=['844127424526680084']),
                simplyplural_id=None
            ).model_dump(mode='json')),
        Example(
            name='Member with Userproxy (With Token Scope)',
            value=MemberModel(
                id=PydanticObjectId('67cab2a55f8b2e7fd3d27d0c'),
                name='steve',
                meta='',
                pronouns='any/all',
                bio='example bio',
                birthday='01/01/2000',
                color=16777065,
                avatar='f873c5cff5608ff4952cfe8b80f1a86c',
                proxy_tags=[
                    MemberModel.ProxyTag(
                        id=PydanticObjectId('67cab28fe3188a0a7807c8fd'),
                        prefix='s;',
                        suffix='',
                        regex=False,
                        case_sensitive=False,
                        avatar=None),
                    MemberModel.ProxyTag(
                        id=PydanticObjectId('67cab28fe3188a0a7807c8fe'),
                        prefix='',
                        suffix='--steve',
                        regex=False,
                        case_sensitive=False,
                        avatar=None)],
                userproxy=MemberModel.UserProxy(
                    bot_id='1297857704512978984',
                    public_key='267b5eb2ebdbe0620160b0cd99630420cce9f1fa271feaf16bee8a20988dc682',
                    token='MTI5Nzg1NzcwNDUxMjk3ODk4NA.GYv_Xp.mM-h1AoambBtN4JApr0NPQT1twx2fo-kuKvH0o',
                    command='st',
                    guilds=['844127424526680084']),
                simplyplural_id=None
            ).model_dump(mode='json')),
        Example(
            name='Member without Userproxy',
            value=MemberModel(
                id=PydanticObjectId('67cab2a55f8b2e7fd3d27d0c'),
                name='steve',
                meta='',
                pronouns='any/all',
                bio='example bio',
                birthday='01/01/2000',
                color=16777065,
                avatar='f873c5cff5608ff4952cfe8b80f1a86c',
                proxy_tags=[
                    MemberModel.ProxyTag(
                        id=PydanticObjectId('67cab28fe3188a0a7807c8fd'),
                        prefix='s;',
                        suffix='',
                        regex=False,
                        case_sensitive=False,
                        avatar=None),
                    MemberModel.ProxyTag(
                        id=PydanticObjectId('67cab28fe3188a0a7807c8fe'),
                        prefix='',
                        suffix='--steve',
                        regex=False,
                        case_sensitive=False,
                        avatar=None)],
                userproxy=None,
                simplyplural_id=None
            ).model_dump(mode='json')
        )
    ]
)

File: core/src/db/member.rs
use std::collections::HashSet;

use bson::oid::ObjectId;
use mongo_document::Document;
use serde::{Deserialize, Serialize};

#[derive(Debug, Document, Serialize, Deserialize)]
#[document(collection = "members")]
pub struct Member {
    #[serde(rename = "_id")]
    pub id: ObjectId,
    pub name: String,
    pub custom_id: String,
    pub pronouns: String,
    pub bio: String,
    pub birthday: String,
    pub color: Option<i32>,
    pub avatar: Option<String>,
    pub proxy_tags: Vec<ProxyTag>,
    pub userproxy: Option<Userproxy>,
    pub simplyplural_id: Option<String>
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProxyTag {
    pub id: ObjectId,
    pub prefix: String,
    pub suffix: String,
    pub regex: bool,
    pub case_sensitive: bool,
    pub avatar: Option<String>
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Userproxy {
    pub bot_id:     u64,
    pub public_key: String,
    pub token:      String,
    pub command:    String,
    pub guilds:     HashSet<u64>
}

// impl Member {
//     pub fn expired(&self) -> bool {
//         (
//             self.ts.to_chrono() +
//             chrono::Duration::minutes(14) +
//             chrono::Duration::seconds(30)
//         ) < chrono::Utc::now()
//     }
// }
